File: tck/java-tck/src/main/java/kalix/tck/model/view/ViewTckModelImpl.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.tck.model.view;

import kalix.javasdk.view.ViewContext;

public class ViewTckModelImpl extends AbstractViewTckModelView {

  public ViewTckModelImpl(ViewContext context) {}

  @Override
  public View.ViewState emptyState() {
    return null;
  }

  @Override
  public UpdateEffect<View.ViewState> processUpdateUnary(View.ViewState state, View.Event event) {
    if (event.hasReturnAsIs()) {
      return effects()
          .updateState(
              View.ViewState.newBuilder().setData(event.getReturnAsIs().getData()).build());
    } else if (event.hasUppercaseThis()) {
      return effects()
          .updateState(
              View.ViewState.newBuilder()
                  .setData(event.getUppercaseThis().getData().toUpperCase())
                  .build());
    } else if (event.hasAppendToExistingState()) {;
      if (state == null) throw new IllegalArgumentException("State was null for " + event);
      return effects()
          .updateState(
              state
                  .toBuilder()
                  .setData(state.getData() + event.getAppendToExistingState().getData())
                  .build());
    } else if (event.hasFail()) {
      throw new RuntimeException("Fail");
    } else if (event.hasIgnore()) {
      return effects().ignore();
    } else {
      throw new RuntimeException("Unexpected event type " + event.getEventCase().name());
    }
  }
}

File: tck/scala-tck/src/main/scala/kalix/tck/model/view/ViewTckModelImpl.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.tck.model.view

import kalix.scalasdk.view.View.UpdateEffect

class ViewTckModelImpl extends AbstractViewTckModelView {

  override def emptyState: ViewState = ViewState.defaultInstance

  override def processUpdateUnary(state: ViewState, event: Event): UpdateEffect[ViewState] =
    event.event match {
      case Event.Event.ReturnAsIs(ReturnAsIsEvent(data, _))       => effects.updateState(ViewState(data))
      case Event.Event.UppercaseThis(UppercaseThisEvent(data, _)) => effects.updateState(ViewState(data.toUpperCase))
      case Event.Event.AppendToExistingState(_) if state eq null =>
        throw new IllegalArgumentException("State was null for " + event)
      case Event.Event.AppendToExistingState(AppendToExistingState(data, _)) =>
        effects.updateState(ViewState(state.data + data))
      case Event.Event.Fail(_)   => effects.error("Fail")
      case Event.Event.Ignore(_) => effects.ignore()
      case Event.Event.Empty     => effects.ignore()
    }
}
